classes/webservice/WebserviceSpecificManagementAttachments.php

```diff
@@ -0,0 +1,300 @@
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ *  @author 	PrestaShop SA <contact@prestashop.com>
+ *  @copyright  2007-2016 PrestaShop SA
+ *  @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *  International Registered Trademark & Property of PrestaShop SA
+ */
+use Symfony\Component\HttpFoundation\Response;
+
+class WebserviceSpecificManagementAttachmentsCore implements WebserviceSpecificManagementInterface
+{
+    /**
+     * @var WebserviceOutputBuilder
+     */
+    protected $objOutput;
+
+    protected $output;
+
+    /**
+     * @var WebserviceRequest
+     */
+    protected $wsObject;
+
+    /**
+     * The configuration parameters of the current resource
+     *
+     * @var array
+     */
+    public $resourceConfiguration;
+
+    protected $attachment_id;
+
+    protected $displayFile;
+
+    /*
+     * ------------------------------------------------
+     * GETTERS & SETTERS
+     * ------------------------------------------------
+     */
+
+    /**
+     * @param WebserviceOutputBuilderCore $obj
+     *
+     * @return WebserviceSpecificManagementInterface
+     */
+    public function setObjectOutput(WebserviceOutputBuilderCore $obj)
+    {
+        $this->objOutput = $obj;
+
+        return $this;
+    }
+
+    public function getObjectOutput()
+    {
+        return $this->objOutput;
+    }
+
+    public function setWsObject(WebserviceRequestCore $obj)
+    {
+        $this->wsObject = $obj;
+
+        return $this;
+    }
+
+    public function getWsObject()
+    {
+        return $this->wsObject;
+    }
+
+    public function getContent()
+    {
+        if ($this->displayFile) {
+            // if displayFile is set, present the file (download)
+            $this->objOutput->setHeaderParams('Content-Type', $this->displayFile['mime']);
+            $this->objOutput->setHeaderParams('Content-Length', $this->displayFile['file_size']);
+            $this->objOutput->setHeaderParams('Content-Disposition', 'attachment; filename="' . utf8_decode($this->displayFile['file_name']) . '"');
+
+            return file_get_contents($this->displayFile['file']);
+        } else {
+            // Emulate non-specific management
+            $this->wsObject->objectSpecificManagement = false;
+            $this->wsObject->returnOutput();
+        }
+    }
+
+    public function manage()
+    {
+        $this->manageAttachments();
+
+        return $this->wsObject->getOutputEnabled();
+    }
+
+    public function manageAttachments()
+    {
+        // Pre configuration...
+        if (isset($this->wsObject->urlSegment)) {
+            for ($i = 1; $i < 6; ++$i) {
```

Comment:
Understanding these urlSegment seems complicated at the first sight. We need more details on their purpose.

Reply:
I need more time to understand what's happened here ^^

Reply:
Could @Perfecticus help you here?

Reply:
Right, that part was inherited from WebserviceSpecificManagementImages (which uses deeper api path).
Looping for 6 segments is excessive as only the [1] index is used for attachments.

The explanation for the mapping can be seen further down "Available cases api/...".
If urlSegment[1] is set to 'file', binary operations are done (file upload/download)
Otherwise default webservice operations are done (read/write Model information using XML/json).

Typical example:
POST to **https://domain.tld/api/attachments/** only creates model information (similar to any other default api), no file information
POST to **https://domain.tld/api/attachments/file** creates an attachment AND uploads a file for it.

Reply:
Thanks for explanation, I add a comment based on what you write :)